api/converters.py

```diff
@@ -0,0 +1,208 @@
+# -*- coding: utf-8 -*-
+# Copyright 2022 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+from typing import Any
+from google.cloud import ndb  # type: ignore
+
+from internals.core_enums import *
+from internals.core_models import Feature
+
+SIMPLE_TYPES = (int, float, bool, dict, str, list)
+
+def to_dict(entity: ndb.Model) -> dict[str, Any]:
+  output = {}
+  for key, prop in entity._properties.items():
+    # Skip obsolete values that are still in our datastore
+    if not hasattr(entity, key):
+      continue
+
+    value = getattr(entity, key)
+
+    if value is None or isinstance(value, SIMPLE_TYPES):
+      output[key] = value
+    elif isinstance(value, datetime.date):
+      # Convert date/datetime to ms-since-epoch ("new Date()").
+      #ms = time.mktime(value.utctimetuple())
+      #ms += getattr(value, 'microseconds', 0) / 1000
+      #output[key] = int(ms)
+      output[key] = str(value)
+    elif isinstance(value, ndb.GeoPt):
+      output[key] = {'lat': value.lat, 'lon': value.lon}
+    elif isinstance(value, ndb.Model):
+      output[key] = value.to_dict()
+    elif isinstance(value, ndb.model.User):
+      output[key] = value.email()
+    else:
+      raise ValueError('cannot encode ' + repr(prop))
+  return output
+
+def del_none(d):
+  """
+  Delete dict keys with None values, and empty lists, recursively.
+  """
+  for key, value in list(d.items()):
+    if value is None or (isinstance(value, list) and len(value) == 0):
+      del d[key]
+    elif isinstance(value, dict):
+      del_none(value)
+  return d
+
+def migrate_views(f: Feature) -> None:
+  """Migrate obsolete values for views on first edit."""
+  if f.ff_views == MIXED_SIGNALS:
+    f.ff_views = NO_PUBLIC_SIGNALS
+  if f.ff_views == PUBLIC_SKEPTICISM:
+    f.ff_views = OPPOSED
+
+  if f.ie_views == MIXED_SIGNALS:
+    f.ie_views = NO_PUBLIC_SIGNALS
+  if f.ie_views == PUBLIC_SKEPTICISM:
+    f.ie_views = OPPOSED
+
+  if f.safari_views == MIXED_SIGNALS:
+    f.safari_views = NO_PUBLIC_SIGNALS
+  if f.safari_views == PUBLIC_SKEPTICISM:
+    f.safari_views = OPPOSED
+
+def feature_to_legacy_json(f: Feature) -> dict[str, Any]:
+  migrate_views(f)
+  d: dict[str, Any] = to_dict(f)
+  is_released = f.impl_status_chrome in RELEASE_IMPL_STATES
+  d['is_released'] = is_released
+
+  standard_maturity_val = f.standard_maturity
+  if (standard_maturity_val == UNSET_STD and
```

Comment:
Can this be deleted now?
#2195 landed back in Sept.

Reply:
Yes! removed.